### ISSUE TITLE
Inconsistent typechecker behaviour of equivalent dataclass usages

### ISSUE THREAD

**Issue by Tomatosoup97** [python/mypy#10849]
**Bug Report**

According to python [docs](https://docs.python.org/3/library/dataclasses.html#module-level-decorators-classes-and-functions) using `@dataclass` and `@dataclass()` should be equivalent, however, based on which one you choose you may get a different result from typechecker - in my example, while using structural subtyping. It occurs iff `show_none_errors = False` is set.

**To Reproduce**

```
# poc.py
from dataclasses import dataclass
from typing import Protocol

@dataclass
class Large:
    x: int
    y: int

@dataclass
class Small(Protocol):
    x: int

def f(x: Small) -> Small:
    return x

f(Large(1, 2))
```

>$ mypy poc.py
>Success: no issues found in 1 source file

now, write dataclass() instead of dataclass

>$ mypy poc.py
>poc.py:10: error: Only concrete class can be given where "Type[Small]" is expected

**Expected Behavior**

Consistent result for both versions - [apparently both should fail](https://github.com/python/mypy/issues/5374#issuecomment-884186452)

**Your Environment**

- Mypy version used: 0.910
- Mypy configuration options from `mypy.ini` (and other config files): `show_none_errors = False`
- Python version used: 3.9.6

**Comment by AlexWaygood** [comment]
I can't reproduce -- for me, mypy [gives the same](https://mypy-play.net/?mypy=latest&python=3.10&flags=strict&gist=77a62b3adcdc047c2f3b6db21db6124d) (false-positive) error for `@dataclass()` and `@dataclass`.

(On mypy 0.941, the latest version available on mypy playground, it emits an additional false-positive error about incompatible `__match_args__`, but that's been fixed in 0.942.)
